pkg/agent/openflow/pipeline.go

```diff
@@ -2923,21 +2923,23 @@ func (f *featureService) gatewaySNATFlows() []binding.Flow {
 				MatchProtocol(ipProtocol).
 				MatchCTStateNew(true).
 				MatchCTStateTrk(true).
-				MatchRegMark(GatewayHairpinRegMark).
+				MatchRegMark(FromGatewayRegMark).
+				MatchRegMark(ToGatewayRegMark).
 				Action().CT(true, ServiceMarkTable.GetNext(), f.dnatCtZones[ipProtocol], f.ctZoneSrcField).
 				LoadToCtMark(ConnSNATCTMark).
 				LoadToCtMark(HairpinCTMark).
 				CTDone().
 				Done(),
 			// This generates the flow to match the first packet of NodePort / LoadBalancer connection initiated through the
-			// Antrea gateway and externalTrafficPolicy of the Service is Cluster. ConnSNATCTMark will be loaded in DNAT
-			// CT zone.
+			// Antrea gateway and externalTrafficPolicy of the Service is Cluster, and the selected Endpoint is on remote
+			// Node, then ConnSNATCTMark will be loaded in DNAT CT zone.
 			ServiceMarkTable.ofTable.BuildFlow(priorityLow).
 				Cookie(cookieID).
 				MatchProtocol(ipProtocol).
 				MatchCTStateNew(true).
 				MatchCTStateTrk(true).
 				MatchRegMark(FromGatewayRegMark).
+				MatchRegMark(ToTunnelRegMark).
```

Comment:
Does this flow require encap mode?

Reply:
Yes, only for noEncap mode, it is a hairpin case.

Reply:
Currently, for matching ToTunnelRegMark, it's only in encap mode; for matching ToUplinkRegMark, it's only in noEncap mode. I prefer to preserve both flows in case that they are all needed at the same time in the future.
How do you think @wenyingd @tnqn @jianjuns ? Thanks.

Reply:
For hybrid mode, these two flows should be both needed. But for noEncap mode, this flow is never hit, so we don't need to install it.

Reply:
Use `NetworkConfig.TrafficEncapMode.SupportsEncap()`?

Reply:
Updated.